core/src/replay_stage.rs

```diff
@@ -199,11 +203,43 @@ impl ReplayStage {
             snapshot_package_sender,
             block_commitment_cache,
             transaction_status_sender,
+            tower_snapshot_path,
+            allow_missing_tower_state,
         } = config;
 
         let (root_bank_sender, root_bank_receiver) = channel();
         trace!("replay stage");
-        let mut tower = Tower::new(&my_pubkey, &vote_account, &bank_forks.read().unwrap());
+        let tower = Tower::reload_from_file(&tower_snapshot_path, &my_pubkey, &vote_account);
+
+        let mut tower = match tower {
+            Ok(tower) => tower,
+            Err(e) => {
+                let bank_forks_lock = bank_forks.read().unwrap();
+                if let Some(account) = bank_forks_lock
+                    .get(bank_forks_lock.root())
+                    .expect("Failed to get root bank")
+                    .get_account(&vote_account)
+                {
+                    if let Some(vote_state) = VoteState::from(&account) {
+                        if !vote_state.votes.is_empty() {
```

Comment:
Hmmm, just because the vote state is empty is no guarantee that this validator hasn't already voted right? This validator could have voted and the votes didn't land in the bank.

Reply:
@carllin the validator can wait `MAX_RECENT_BLOCKHASHES` before it starts voting to alleviate this case, correct?